### PR TITLE
Squash all lints tied to foreign macros by default

### LINKED ISSUE #48855: Add macro checks to lints
From https://github.com/rust-lang/rust/issues/48385

Many of our lints trigger even when the code is within a macro defined by an external crate. They should include macro checks.

Clippy has [a function](https://github.com/rust-lang-nursery/rust-clippy/blob/a7407f9acc2e216dee23fd5e4fc2ab9b01e4d77b/clippy_lints/src/utils/mod.rs#L67-L95) for this already.

We should include this in librustc::lint somewhere, and use it for various lints. It should _not_ be used for future compat lints.

One possible thing to do here is to work it into the lint framework itself, so that non-future-compat lints which are triggered on macro spans will not actually be emitted.

We should still have the helper function since some lints may involve multiple important spans so they need the ability to check them all.

Ideally there would be a way to turn this off, though -- in case you want to find issues in your macros. It's a tricky problem.

Willing to mentor the basic issue of importing the macro checks and applying them in most lints (or working it into the framework).

### LINKED ISSUE #52483: False positive for 2018 unused extern crate in global allocator
Possibly macro related, not sure.

https://play.rust-lang.org/?gist=7c9846d22f310efa6ab4bec086d61d4e&version=nightly&mode=debug&edition=2018

```rust
#![deny(rust_2018_idioms)]

use alloc_system;

#[global_allocator]
static ALLOC: alloc_system::System = alloc_system::System;

fn main() {}
```

```
error: `extern crate` is not idiomatic in the new edition
 --> src/main.rs:6:1
  |
6 | static ALLOC: alloc_system::System = alloc_system::System;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: convert it to a `use`
  |
note: lint level defined here
 --> src/main.rs:1:9
  |
1 | #![deny(rust_2018_idioms)]
  |         ^^^^^^^^^^^^^^^^
  = note: #[deny(unused_extern_crates)] implied by #[deny(rust_2018_idioms)]
```

cc @alexcrichton

### PR DESCRIPTION
**Pull request by alexcrichton**
This PR is a continuation of https://github.com/rust-lang/rust/pull/49755 (thanks for the initial jump-start @Dylan-DPC!) and is targeted at solving https://github.com/rust-lang/rust/issues/48855. This change updates the lint infrastructure to, by default, ignore all lints emitted for code that originates in a foreign macro. For example if `println!("...")` injects some idiomatic warnings these are all ignored by default. The rationale here is that for almost all lints there's no action that can be taken if the code originates from a foreign lint.

Closes #48855
Closes #52483 
Closes #52479

### CODE REVIEW

**Review by Manishearth — CHANGES_REQUESTED** [review]
Looks good so far, but it should _always_ output warnings for future incompatibility lints. Folks should know if their code will break, even if it's not something they can fix in their own crate.

**Comment by alexcrichton** [comment]
@Manishearth indeed! That's why this is in the `else` clause where the previous has to do with future incompatibility (this is also mentioned in the comment added)

**Comment by oli-obk** [comment]
Ideally we'd still emit lints if the macro originates from a crate within the workspace. But that probably requires significant changes to tell us which crates belong to the same workspace

File: src/librustc/lint/mod.rs [hunk 1]
```diff
@@ -669,3 +685,33 @@ impl<'a, 'tcx> intravisit::Visitor<'tcx> for LintLevelMapBuilder<'a, 'tcx> {
 pub fn provide(providers: &mut Providers) {
     providers.lint_levels = lint_levels;
 }
+
+/// Returns whether `span` originates in a foreign crate's external macro.
+///
+/// This is used to test whether a lint should be entirely aborted above.
+fn in_external_macro(sess: &Session, span: Span) -> bool {
```
Comment (oli-obk) [inline src/librustc/lint/mod.rs]: Can you make this public so we can reuse it from clippy?

**Review by Manishearth — APPROVED** [review]
lol missed that (was reviewing from phone)

r=me, with the function public

File: src/librustc/lint/mod.rs (hunk 1 above)
Comment (alexcrichton) [reply src/librustc/lint/mod.rs]: Certainly!

**Comment by alexcrichton** [comment]
@oli-obk I agree! I also agree though that unfortunately there's no great way to know that in the compiler right now :(

@bors: r=Manishearth

**Comment by Manishearth** [comment]
This ends up breaking a bunch of clippy lints:

 - Sometimes we lint on an expression when not caring about the contents, it should be okay for that to be a macro -- e.g. if the expression is `println!("foo")`, the lint gets hidden, even though the lint is about the expression as a whole and not the macro origins
 - There should be an opt out, we have some lints that lint `#[derive()]`s
 - There's some funkyness around compiler desugarings, we're having  `for i in 1..3` where the lint is on the range getting squashed

overall ISTM it might be better to make this a per-lint thing like how clippy does. Either way, clippy is going to opt out I think.

**Comment by alexcrichton** [comment]
@Manishearth should [this check](https://github.com/rust-lang/rust/blob/878dd0b5e19e086e608351f33bf12e3625425f20/src/librustc/lint/mod.rs#L586-L587) perhaps become a runtime check against a set stored in the session or lint context? That way Clippy would be able to register lints as "please issue a warning despite the foreign-macro-ness"

**Comment by oli-obk** [comment]
I have made it a runtime check and added the two rustc exceptions as a proof of concept: https://github.com/rust-lang/rust/pull/52562/files#diff-837439bcaa26732bb48bcbb0c60716ccL586

**Pull request merged**